File: ios/Nufo/Data/Parsers.swift
import Foundation

typealias JSON = [String: Any]

private extension Dictionary where Key == String, Value == Any {
    func str(_ k: String) -> String? {
        guard let s = self[k] as? String else { return nil }
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
    func num(_ k: String) -> Double? {
        if let d = self[k] as? Double { return d }
        if let i = self[k] as? Int { return Double(i) }
        if let s = self[k] as? String { return Double(s) }
        return nil
    }
    func int(_ k: String) -> Int? { num(k).map { Int($0) } }
    func tags(_ k: String) -> [String] { (self[k] as? [String]) ?? [] }
}

/// "en:sesame-seeds" -> "Sesame seeds". Last resort when no translated name exists.
func prettyTag(_ tag: String) -> String {
    let raw = (tag.split(separator: ":", maxSplits: 1).last.map(String.init) ?? tag).replacingOccurrences(of: "-", with: " ")
    return raw.prefix(1).uppercased() + raw.dropFirst()
}

private func grade(_ v: String?) -> String? {
    guard let g = v?.lowercased(), g.count == 1, "abcde".contains(g) else { return nil }
    return g
}

/// OFF sometimes HTML-escapes names ("Φέτα &quot;Ελαφρύ&quot;").
private func unescape(_ s: String) -> String {
    s.replacingOccurrences(of: "&quot;", with: "\"").replacingOccurrences(of: "&amp;", with: "&")
        .replacingOccurrences(of: "&#39;", with: "'").replacingOccurrences(of: "&apos;", with: "'")
}

/// Drops the EU "℮" estimated-quantity mark, which OFF often stores as a trailing " e" ("400 g e").
func stripEstimatedMark(_ q: String) -> String {
    q.replacingOccurrences(of: #"\s*(℮|\be)\s*$"#, with: "", options: .regularExpression).trimmingCharacters(in: .whitespaces)
}

/// Energy in kcal, rejecting impossible values: nothing exceeds 900 kcal per 100 g (pure fat).
/// A kcal field above that is usually kJ typed into the wrong box, so the kJ field (/ 4.184) is used
/// instead, or nothing. `perGrams` is nil for per-serving values, which have no fixed ceiling.
func kcal(_ kcal: Double?, _ kj: Double?, perGrams: Double?) -> Double? {
    func ok(_ v: Double) -> Bool { v >= 0 && (perGrams == nil || v <= 9 * perGrams!) }
    if let k = kcal, ok(k) { return k }
    if let j = kj.map({ $0 / 4.184 }), ok(j) { return j }
    return nil
}

enum OffParser {
    static let source = "Open Food Facts"
    static let fields = "code,product_name,product_name_el,product_name_en,brands,quantity,serving_size,serving_quantity,nutriments,nutriscore_grade,nova_group,ecoscore_grade,ingredients_text,ingredients_text_el,ingredients_text_en,allergens_tags,additives_tags,image_url,categories_tags,labels_tags,countries_tags,completeness,last_modified_t"
    static let searchFields = "code,product_name,product_name_el,product_name_en,brands,image_url,nutriscore_grade,nova_group,nutriments,countries_tags"

    private static func nutrition(_ n: JSON, _ suffix: String) -> Nutrition? {
        func v(_ k: String) -> Double? { n.num("\(k)_\(suffix)") }
        var out = Nutrition(
            calories: kcal(v("energy-kcal"), v("energy"), perGrams: suffix == "100g" ? 100 : nil),
            protein: v("proteins"), carbs: v("carbohydrates"), fat: v("fat"), saturatedFat: v("saturated-fat"),
            fiber: v("fiber"), sugar: v("sugars"), sodium: v("sodium"), salt: v("salt")
        )
        for id in Micronutrients.vitaminIds { if let x = v(id) { out.vitamins[id] = x } }
        for id in Micronutrients.mineralIds { if let x = v(id) { out.minerals[id] = x } }
        return out.isEmpty ? nil : out
    }

    /// Text in the app language when OFF has it, else in the product's main language.
    private static func localized(_ p: JSON, _ field: String, _ lang: String) -> String? {
        (p.str("\(field)_\(lang)") ?? p.str(field) ?? p.str("\(field)_en") ?? p.str("\(field)_el")).map(unescape)
    }

    /// nil when OFF reports the product as missing (status != 1).
    static func parseProductResponse(_ root: JSON, barcode: String, lang: String = "en") -> Product? {
        guard root.int("status") == 1, let p = root["product"] as? JSON else { return nil }
        return parseProduct(p, barcode: barcode, lang: lang)
    }

    static func parseProduct(_ p: JSON, barcode: String?, lang: String = "en") -> Product {
        let nutriments = p["nutriments"] as? JSON ?? [:]
        let per100 = nutrition(nutriments, "100g") ?? Nutrition()
        let servingGrams = p.num("serving_quantity").flatMap { $0 > 0 ? $0 : nil }
        let perServing = nutrition(nutriments, "serving") ?? (per100.isEmpty ? nil : servingGrams.map { per100.scaled($0 / 100) })
        let nutri = grade(p.str("nutriscore_grade"))
        let nova = p.int("nova_group").flatMap { (1...4).contains($0) ? $0 : nil }
        return Product(
            barcode: barcode ?? p.str("code"),
            name: localized(p, "product_name", lang) ?? "",
            brand: p.str("brands")?.split(separator: ",").first.map { $0.trimmingCharacters(in: .whitespaces) },
            quantity: p.str("quantity").map(stripEstimatedMark).flatMap { $0.isEmpty ? nil : $0 }, servingSize: p.str("serving_size"), servingGrams: servingGrams,
            imageUrl: p.str("image_url"),
            ingredientsText: localized(p, "ingredients_text", lang)?.replacingOccurrences(of: "_", with: ""), // OFF wraps allergens in _underscores_
            allergenTags: p.tags("allergens_tags"), additiveTags: p.tags("additives_tags"),
            categoryTags: p.tags("categories_tags"), labelTags: p.tags("labels_tags"),
            nutritionPer100g: per100, nutritionPerServing: perServing,
            nutriscoreGrade: nutri, novaGroup: nova, ecoscoreGrade: grade(p.str("ecoscore_grade")),
            nufoScore: Scoring.nufoScore(per100, nutriscore: nutri, nova: nova).score,
            source: source, lastUpdated: Int64(p.num("last_modified_t") ?? 0) * 1000,
            completeness: p.num("completeness").map { min(1, max(0, $0)) },
            soldInGreece: p.tags("countries_tags").contains("en:greece")
        )
    }

    /// Legacy search.pl returns "products"; Search-a-licious returns "hits" with brands as an array.
    static func parseSearch(_ root: JSON, lang: String = "en") -> [SearchHit] {
        let items = (root["hits"] as? [JSON]) ?? (root["products"] as? [JSON]) ?? []
        return items.compactMap { p in
            guard let name = localized(p, "product_name", lang) else { return nil }
            let brand = (p["brands"] as? [String])?.first?.trimmingCharacters(in: .whitespaces)
                ?? p.str("brands")?.split(separator: ",").first.map { $0.trimmingCharacters(in: .whitespaces) }
            let n = p["nutriments"] as? JSON
            return SearchHit(
                name: name, brand: brand, imageUrl: p.str("image_url"),
                nutriscoreGrade: grade(p.str("nutriscore_grade")),
                novaGroup: p.int("nova_group").flatMap { (1...4).contains($0) ? $0 : nil },
                caloriesPer100g: n.flatMap { kcal($0.num("energy-kcal_100g"), $0.num("energy_100g"), perGrams: 100) },
                source: source, barcode: p.str("code"),
                soldInGreece: p.tags("countries_tags").contains("en:greece")
            )
        }
    }

    /// Taxonomy response `{"en:milk":{"name":{"el":"γάλα"}}}` -> tag id to translated name.
    static func parseTaxonomy(_ root: JSON, lang: String) -> [String: String] {
        var out: [String: String] = [:]
        for (tag, v) in root {
            if let name = ((v as? JSON)?["name"] as? JSON)?.str(lang) { out[tag] = name.prefix(1).uppercased() + name.dropFirst() }
        }
        return out
    }
}

enum UsdaParser {
    static let source = "USDA FoodData Central"
    private static let mg = 0.001, ug = 0.000001

    static func parseSearch(_ root: JSON) -> [SearchHit] {
        ((root["foods"] as? [JSON]) ?? []).compactMap { f in
            guard let p = parseFood(f) else { return nil }
            return SearchHit(name: p.name, brand: p.brand, imageUrl: nil, nutriscoreGrade: nil, novaGroup: nil,
                             caloriesPer100g: p.nutritionPer100g.calories, source: source, barcode: p.barcode, product: p)
        }
    }

    /// FDC search values are per 100 g, keyed by legacy nutrient number.
    static func parseFood(_ f: JSON) -> Product? {
        guard let name = f.str("description") else { return nil }
        var nutrients: [String: Double] = [:]
        for o in (f["foodNutrients"] as? [JSON]) ?? [] {
            if let n = o.str("nutrientNumber"), let v = o.num("value") { nutrients[n] = v }
        }
        func g(_ n: String, _ factor: Double = 1) -> Double? { nutrients[n].map { $0 * factor } }
        let sodium = g("307", mg)
        var per100 = Nutrition(
            calories: g("208") ?? g("957") ?? g("958"), protein: g("203"), carbs: g("205"), fat: g("204"),
            saturatedFat: g("606"), fiber: g("291"), sugar: g("269"), sodium: sodium,
            salt: sodium.map { $0 * 2.5 } // EU labelling convention: salt = sodium x 2.5
        )
        for (n, id, f) in [("320", "vitamin-a", ug), ("401", "vitamin-c", mg), ("328", "vitamin-d", ug),
                           ("323", "vitamin-e", mg), ("415", "vitamin-b6", mg), ("418", "vitamin-b12", ug)] {
            if let v = g(n, f) { per100.vitamins[id] = v }
        }
        for (n, id) in [("301", "calcium"), ("303", "iron"), ("304", "magnesium"), ("306", "potassium"), ("309", "zinc")] {
            if let v = g(n, mg) { per100.minerals[id] = v }
        }
        let unit = f.str("servingSizeUnit")?.lowercased()
        let servingGrams = (unit == "g" || unit == "grm") ? f.num("servingSize") : nil
        var updated: Int64 = 0
        if let d = f.str("publishedDate") ?? f.str("modifiedDate") {
            let fmt = DateFormatter(); fmt.dateFormat = "yyyy-MM-dd"; fmt.timeZone = TimeZone(identifier: "UTC")
            if let date = fmt.date(from: String(d.prefix(10))) { updated = Int64(date.timeIntervalSince1970 * 1000) }
        }
        return Product(
            barcode: f.str("gtinUpc"), name: name.lowercased().prefix(1).uppercased() + name.lowercased().dropFirst(),
            brand: f.str("brandOwner") ?? f.str("brandName"), quantity: nil,
            servingSize: servingGrams.map { "\(Int($0)) g" } ?? f.str("householdServingFullText"), servingGrams: servingGrams,
            imageUrl: nil, ingredientsText: f.str("ingredients"),
            categoryTags: [],
            nutritionPer100g: per100, nutritionPerServing: servingGrams.map { per100.scaled($0 / 100) },
            nutriscoreGrade: nil, novaGroup: nil, ecoscoreGrade: nil,
            nufoScore: Scoring.nufoScore(per100, nutriscore: nil, nova: nil).score,
            source: source, lastUpdated: updated
        )
    }
}
enum PricesParser {
    static let source = "Open Prices"

    /// Latest reports, Greek stores first; at most `limit`.
    static func parse(_ root: JSON, limit: Int = 3) -> [PriceReport] {
        let items = (root["items"] as? [JSON]) ?? []
        let reports: [PriceReport] = items.compactMap { o in
            guard let price = o.num("price"), price > 0, let currency = o.str("currency"), let date = o.str("date") else { return nil }
            let loc = o["location"] as? JSON
            return PriceReport(price: price, currency: currency, date: date,
                               store: loc?.str("osm_name"), city: loc?.str("osm_address_city"), country: loc?.str("osm_address_country"))
        }
        return Array(reports.enumerated().sorted { a, b in
            a.element.inGreece != b.element.inGreece ? a.element.inGreece : a.offset < b.offset
        }.map(\.element).prefix(limit))
    }
}

enum UpcItemDbParser {
    static let source = "UPCitemdb"

    static func parse(_ root: JSON) -> Identified? {
        guard let item = (root["items"] as? [JSON])?.first, let title = item.str("title") else { return nil }
        // Their image links point at third-party shops; only https ones are safe to load.
        return Identified(name: title, brand: item.str("brand"),
                          imageUrl: item.tags("images").first { $0.hasPrefix("https://") }, source: source)
    }
}

File: android/app/src/main/java/com/nufo/app/data/Parsers.kt
package com.nufo.app.data

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import java.time.LocalDate
import java.time.ZoneOffset

private fun JsonObject.str(key: String) = (this[key] as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { it.isNotEmpty() }
private fun JsonObject.num(key: String) = (this[key] as? JsonPrimitive)?.let { it.doubleOrNull ?: it.contentOrNull?.toDoubleOrNull() }
private fun JsonObject.int(key: String) = (this[key] as? JsonPrimitive)?.let { it.intOrNull ?: it.contentOrNull?.toIntOrNull() }
private fun JsonObject.tags(key: String) = (this[key] as? JsonArray)?.mapNotNull { (it as? JsonPrimitive)?.contentOrNull } ?: emptyList()

/** "en:sesame-seeds" -> "Sesame seeds". Last resort when no translated name exists. */
fun prettyTag(tag: String): String = tag.substringAfter(':').replace('-', ' ').replaceFirstChar { it.uppercase() }

/**
 * Energy in kcal, rejecting impossible values: nothing exceeds 900 kcal per 100 g (pure fat).
 * A kcal field above that is usually kJ typed into the wrong box, so the kJ field (÷ 4.184) is
 * used instead, or nothing. [perGrams] is null for per-serving values, which have no fixed ceiling.
 */
internal fun kcal(kcal: Double?, kj: Double?, perGrams: Double?): Double? {
    fun ok(v: Double) = v >= 0 && (perGrams == null || v <= 9.0 * perGrams)
    return kcal?.takeIf(::ok) ?: kj?.div(4.184)?.takeIf(::ok)
}

private fun grade(value: String?) = value?.lowercase()?.takeIf { it.length == 1 && it[0] in 'a'..'e' }

/** OFF sometimes HTML-escapes names ("Φέτα &quot;Ελαφρύ&quot;"). */
private fun unescape(s: String) = s.replace("&quot;", "\"").replace("&amp;", "&").replace("&#39;", "'").replace("&apos;", "'")

object OffParser {
    const val SOURCE = "Open Food Facts"
    // EU "℮" estimated-quantity mark, which OFF often stores as a trailing " e".
    private val ESTIMATED_MARK = Regex("""\s*(℮|\be)\s*$""")

    const val FIELDS = "code,product_name,product_name_el,product_name_en,brands,quantity,serving_size,serving_quantity," +
        "nutriments,nutriscore_grade,nova_group,ecoscore_grade,ingredients_text,ingredients_text_el,ingredients_text_en," +
        "allergens_tags,additives_tags,image_url,categories_tags,labels_tags,countries_tags,completeness,last_modified_t"
    const val SEARCH_FIELDS = "code,product_name,product_name_el,product_name_en,brands,image_url,nutriscore_grade,nova_group,nutriments,countries_tags"

    private fun nutrition(n: JsonObject, suffix: String): Nutrition? {
        fun v(key: String) = n.num("${key}_$suffix")
        val result = Nutrition(
            calories = kcal(v("energy-kcal"), v("energy"), if (suffix == "100g") 100.0 else null),
            protein = v("proteins"),
            carbs = v("carbohydrates"),
            fat = v("fat"),
            saturatedFat = v("saturated-fat"),
            fiber = v("fiber"),
            sugar = v("sugars"),
            sodium = v("sodium"),
            salt = v("salt"),
            vitamins = Micronutrients.vitamins.keys.mapNotNull { k -> v(k)?.let { k to it } }.toMap(),
            minerals = Micronutrients.minerals.keys.mapNotNull { k -> v(k)?.let { k to it } }.toMap(),
        )
        return result.takeUnless { it == Nutrition() }
    }

    /** Name in the app language when OFF has one, else the name in the product's main language. */
    private fun localized(p: JsonObject, field: String, lang: String) =
        (p.str("${field}_$lang") ?: p.str(field) ?: p.str("${field}_en") ?: p.str("${field}_el"))?.let(::unescape)

    /** Returns null when OFF reports the product as missing (status != 1). */
    fun parseProductResponse(root: JsonObject, barcode: String, lang: String = "en"): Product? {
        if (root.int("status") != 1) return null
        val product = root["product"] as? JsonObject ?: return null
        return parseProduct(product, barcode, lang)
    }

    fun parseProduct(p: JsonObject, barcode: String?, lang: String = "en"): Product {
        val nutriments = p["nutriments"] as? JsonObject ?: JsonObject(emptyMap())
        val per100 = nutrition(nutriments, "100g") ?: Nutrition()
        val servingGrams = p.num("serving_quantity")?.takeIf { it > 0 }
        val perServing = nutrition(nutriments, "serving")
            ?: servingGrams?.takeIf { per100 != Nutrition() }?.let { per100.scaled(it / 100) }
        val nutri = grade(p.str("nutriscore_grade"))
        val nova = p.int("nova_group")?.takeIf { it in 1..4 }
        return Product(
            barcode = barcode ?: p.str("code"),
            name = localized(p, "product_name", lang) ?: "",
            brand = p.str("brands")?.split(',')?.first()?.trim(),
            quantity = p.str("quantity")?.replace(ESTIMATED_MARK, "")?.trim()?.ifEmpty { null },
            servingSize = p.str("serving_size"),
            servingGrams = servingGrams,
            imageUrl = p.str("image_url"),
            ingredientsText = localized(p, "ingredients_text", lang)?.replace("_", ""), // OFF wraps allergens in _underscores_
            allergenTags = p.tags("allergens_tags"),
            additiveTags = p.tags("additives_tags"),
            categoryTags = p.tags("categories_tags"),
            labelTags = p.tags("labels_tags"),
            nutritionPer100g = per100,
            nutritionPerServing = perServing,
            nutriscoreGrade = nutri,
            novaGroup = nova,
            ecoscoreGrade = grade(p.str("ecoscore_grade")),
            nufoScore = Scoring.nufoScore(per100, nutri, nova).score,
            source = SOURCE,
            lastUpdated = (p.num("last_modified_t")?.toLong() ?: 0L) * 1000,
            completeness = p.num("completeness")?.coerceIn(0.0, 1.0),
            soldInGreece = "en:greece" in p.tags("countries_tags"),
        )
    }

    /** Legacy search.pl returns "products"; Search-a-licious returns "hits" with brands as an array. */
    fun parseSearch(root: JsonObject, lang: String = "en"): List<SearchHit> =
        ((root["hits"] ?: root["products"]) as? JsonArray).orEmpty().mapNotNull { el ->
            val p = el as? JsonObject ?: return@mapNotNull null
            val name = localized(p, "product_name", lang) ?: return@mapNotNull null
            SearchHit(
                name = name,
                brand = p.tags("brands").firstOrNull()?.trim() ?: p.str("brands")?.split(',')?.first()?.trim(),
                imageUrl = p.str("image_url"),
                nutriscoreGrade = grade(p.str("nutriscore_grade")),
                novaGroup = p.int("nova_group")?.takeIf { it in 1..4 },
                caloriesPer100g = (p["nutriments"] as? JsonObject)?.let { kcal(it.num("energy-kcal_100g"), it.num("energy_100g"), 100.0) },
                source = SOURCE,
                barcode = p.str("code"),
                soldInGreece = "en:greece" in p.tags("countries_tags"),
            )
        }

    /** Taxonomy response `{"en:milk":{"name":{"el":"γάλα"}}}` -> tag id to translated name. */
    fun parseTaxonomy(root: JsonObject, lang: String): Map<String, String> =
        root.mapNotNull { (tag, v) ->
            val name = ((v as? JsonObject)?.get("name") as? JsonObject)?.str(lang) ?: return@mapNotNull null
            tag to name.replaceFirstChar { it.uppercase() }
        }.toMap()
}

object UsdaParser {
    const val SOURCE = "USDA FoodData Central"
    private const val MG = 0.001
    private const val UG = 0.000001

    fun parseSearch(root: JsonObject): List<SearchHit> =
        (root["foods"] as? JsonArray).orEmpty().mapNotNull { el ->
            val product = (el as? JsonObject)?.let(::parseFood) ?: return@mapNotNull null
            SearchHit(
                name = product.name, brand = product.brand, imageUrl = null,
                nutriscoreGrade = null, novaGroup = null,
                caloriesPer100g = product.nutritionPer100g.calories,
                source = SOURCE, barcode = product.barcode, product = product,
            )
        }

    /** Values from FDC search are per 100 g; nutrients are keyed by legacy nutrient number. */
    fun parseFood(f: JsonObject): Product? {
        val name = f.str("description") ?: return null
        val nutrients: Map<String, Double> = (f["foodNutrients"] as? JsonArray).orEmpty().mapNotNull { e ->
            val o = e as? JsonObject ?: return@mapNotNull null
            val number = o.str("nutrientNumber") ?: return@mapNotNull null
            val value = o.num("value") ?: return@mapNotNull null
            number to value
        }.toMap()
        fun g(n: String, factor: Double = 1.0) = nutrients[n]?.times(factor)
        val sodium = g("307", MG)
        val per100 = Nutrition(
            calories = g("208") ?: g("957") ?: g("958"),
            protein = g("203"), carbs = g("205"), fat = g("204"), saturatedFat = g("606"),
            fiber = g("291"), sugar = g("269"),
            sodium = sodium,
            salt = sodium?.times(2.5), // EU labelling convention: salt = sodium x 2.5
            vitamins = listOfNotNull(
                g("320", UG)?.let { "vitamin-a" to it }, g("401", MG)?.let { "vitamin-c" to it },
                g("328", UG)?.let { "vitamin-d" to it }, g("323", MG)?.let { "vitamin-e" to it },
                g("415", MG)?.let { "vitamin-b6" to it }, g("418", UG)?.let { "vitamin-b12" to it },
            ).toMap(),
            minerals = listOfNotNull(
                g("301", MG)?.let { "calcium" to it }, g("303", MG)?.let { "iron" to it },
                g("304", MG)?.let { "magnesium" to it }, g("306", MG)?.let { "potassium" to it },
                g("309", MG)?.let { "zinc" to it },
            ).toMap(),
        )
        val servingGrams = f.num("servingSize")?.takeIf { f.str("servingSizeUnit")?.lowercase() in setOf("g", "grm") }
        val date = f.str("publishedDate") ?: f.str("modifiedDate")
        return Product(
            barcode = f.str("gtinUpc"),
            name = name.lowercase().replaceFirstChar { it.uppercase() },
            brand = f.str("brandOwner") ?: f.str("brandName"),
            quantity = null,
            servingSize = servingGrams?.let { "${it.toInt()} g" } ?: f.str("householdServingFullText"),
            servingGrams = servingGrams,
            imageUrl = null,
            ingredientsText = f.str("ingredients"),
            nutritionPer100g = per100,
            nutritionPerServing = servingGrams?.let { per100.scaled(it / 100) },
            nutriscoreGrade = null, novaGroup = null, ecoscoreGrade = null,
            nufoScore = Scoring.nufoScore(per100, null, null).score,
            source = SOURCE,
            lastUpdated = date?.let {
                runCatching { LocalDate.parse(it.take(10)).atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli() }.getOrNull()
            } ?: 0L,
        )
    }
}

object PricesParser {
    const val SOURCE = "Open Prices"

    /** Latest reports, Greek stores first; at most [limit]. */
    fun parse(root: JsonObject, limit: Int = 3): List<PriceReport> =
        (root["items"] as? JsonArray).orEmpty().mapNotNull { el ->
            val o = el as? JsonObject ?: return@mapNotNull null
            val price = o.num("price")?.takeIf { it > 0 } ?: return@mapNotNull null
            val loc = o["location"] as? JsonObject
            PriceReport(
                price = price,
                currency = o.str("currency") ?: return@mapNotNull null,
                date = o.str("date") ?: return@mapNotNull null,
                store = loc?.str("osm_name"),
                city = loc?.str("osm_address_city"),
                country = loc?.str("osm_address_country"),
            )
        }.sortedByDescending { it.inGreece }.take(limit)
}

object UpcItemDbParser {
    const val SOURCE = "UPCitemdb"

    fun parse(root: JsonObject): Identified? {
        val item = (root["items"] as? JsonArray)?.firstOrNull() as? JsonObject ?: return null
        val title = item.str("title") ?: return null
        return Identified(
            name = title,
            brand = item.str("brand"),
            // Their image links point at third-party shops; only https ones are safe to load.
            imageUrl = item.tags("images").firstOrNull { it.startsWith("https://") },
            source = SOURCE,
        )
    }
}